### PR TITLE
Skip writing if nothing changed

### PR DESCRIPTION
**Pull request by pepibumur**
Resolves https://github.com/xcodeswift/xcproj/issues/128

### Short description 📝
As described in the issue the purpose of this PR is to update the write logic to skip writing if nothing changed. This prevents Xcode from reloading the project unnecessarily.

### Solution 📦
Update implementation and add some tests.

### Implementation 👩‍💻👨‍💻
- [x] PBXProj
- [x] XCWorkspaceData
- [x] XCScheme

### GIF
gif

---
This change is [Reviewable](https://reviewable.io/reviews/xcodeswift/xcproj/180)

### CODE REVIEW

**Comment by keith** [comment]
One thing to note is that for anyone using CocoaPods the XML will never be the same since CocoaPods adds some references and build phases.

File: Fixtures/iOS/Project.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -7,7 +7,7 @@
 	objects = {
 
 /* Begin PBXBuildFile section */
-		0403345F1F8F57CB0079CCF5 /* Model.xcdatamodeld in Sources */ = {isa = PBXBuildFile; fileRef = 0403345D1F8F57CB0079CCF5 /* Model.xcdatamodeld */; };
+		0403345F1F8F57CB0079CCF5 = {isa = PBXBuildFile; fileRef = 0403345D1F8F57CB0079CCF5; };
```
Comment (toshi0383) [inline Fixtures/iOS/Project.xcodeproj/project.pbxproj]: Why are these changes necessary?

**Comment by toshi0383** [comment]
> One thing to note is that for anyone using CocoaPods the XML will never be the same since CocoaPods adds some references and build phases.

Yea CocoaPods even updates xcodeproj even when Podfile only contains command-line tools like Sourcery. 🤷‍♂️

**Comment by rahul-malik** [comment]
@yonaskolb - Do you think XcodeGen could leverage this change to update a project vs delete / create a new one?

File: Fixtures/iOS/Project.xcodeproj/project.pbxproj (hunk 1 above)
Comment (pepibumur) [reply Fixtures/iOS/Project.xcodeproj/project.pbxproj]: Nop, I've reverted the change. Good catch 👍

**Comment by yonaskolb** [comment]
I’m actually going to re-implement this in XcodeGen anyway, to have some oversight if anything changed or not, for logging purposes.
I’m actually not 100% convinced this should live in xcproj. Feels weird to have an api do something only in certain cases. Maybe someone wants to rewrite the file for updated permissions or metadata like modified. 
At the very least there should be some way to find out that a file hasn’t actually been written.

**Comment by pepibumur** [comment]
> Maybe someone wants to rewrite the file for updated permissions or metadata like modified.

Isn't modifying permissions or metadata something that you would do without xcproj?

**Comment by pepibumur** [comment]
I'll close this PR leaving up to the developers using xcproj to decide whether they want to override the project all the times or check if anything changed before saving.
